test/test_testing.py

```diff
@@ -3337,6 +3337,45 @@ def test_get_pebble_methods(self):
         client = backend.get_pebble('/custom/socket/path')
         self.assertIsInstance(client, _TestingPebbleClient)
 
+    def test_reboot(self):
+        class RebootingCharm(ops.CharmBase):
+            def __init__(self, framework: ops.Framework):
+                super().__init__(framework)
+                self.framework.observe(self.on.install, self._reboot_now)
+                self.framework.observe(self.on.remove, self._reboot)
+
+            def _reboot_now(self, event: ops.InstallEvent):
+                self.unit.reboot(now=True)
+
+            def _reboot(self, event: ops.RemoveEvent):
+                self.unit.reboot()
+
+        harness = ops.testing.Harness(RebootingCharm, meta='''
+            name: test-app
+            ''')
+        self.addCleanup(harness.cleanup)
+        self.assertIsNone(harness.last_rebooted)
+        backend = harness._backend
+        backend.reboot()
+        reboot1 = harness.last_rebooted
+        assert reboot1 is not None
+        self.assertTrue(reboot1)
+        with self.assertRaises(ops.testing.RebootNow):
+            backend.reboot(now=True)
+        reboot2 = harness.last_rebooted
+        assert reboot2 is not None
+        self.assertGreater(reboot2, reboot1)
```

Comment:
What is the granularity on datetime.now()? Is it guaranteed to have ticked forward since we called it? (if it is nanoseconds probably yes, but if it is milliseconds I might argue not).
I suppose the test is passing, I do wonder if we would want `>=` vs `>`

Here is an oddity:
```
import datetime
n1 = datetime.datetime.now()
try:
  for i in range(100000):
    n2 = datetime.datetime.now()
    if n2 <= n1:
      raise RuntimeError("n2 did not progress")
    n1 = n2
except RuntimeError as e:
  print(e)
  print(n1, n2)
  print(n2-n1)
```

The loop reliably fails

Anyway, I did get:
```
$ python3 ./dt_test.py
n2 did not progress
2023-10-17 16:11:12.484913 2023-10-17 16:11:12.484913
0:00:00
```

Obviously it depends on lots of factors and just assignment doesn't take a lot of time
```
>>> mindt = None
>>> for i in range(1000000):
...   n2 = datetime.datetime.now()
...   dt = n2 - n1
...   if dt != datetime.timedelta(0):
...     if mindt is None or dt < mindt:
...       mindt = dt
...   n1 = n2
...
>>> mindt
datetime.timedelta(microseconds=1)
```

So datetime does only have a resolution of 1 microsecond, which is pretty fast, but you can, indeed do some python work before it ticks over.

Reply:
> What is the granularity on datetime.now()?

The smallest unit in `datetime`s is microseconds, and all the units are ints, so it's one microsecond (at best).

> Is it guaranteed to have ticked forward since we called it?

No. I don't think there is any clock in Python that will do that. The closest would be time's `perf_counter_ns`.

What about `reboot_count`? The time isn't particularly important, but I do want it to be more than a Boolean.